dune/iga/nurbstrimfunctionality.hh

```diff
@@ -0,0 +1,624 @@
+//
+// Created by Henri on 14.03.2023.
+//
+
+#pragma once
+
+namespace Dune::IGA::Impl::Trim {
+
+  // The following numbering is used for edges and nodes:
+  /*
+   * 3 -- 2 -- 2
+   * |         |
+   * 3         1
+   * |         |
+   * 0 -- 0 -- 1
+   */
+
+  constexpr int clipperPrecision = 8;
+  constexpr int pathSamples      = 200;
+  constexpr double tolerance     = 1e-8;
+  // constexpr double epsilonPrecision = double(16) * std::numeric_limits<double>::epsilon();
+
+  using ctype        = double;
+  using ClipperPoint = Clipper2Lib::Point<ctype>;
+  using Point        = FieldVector<ctype, 2>;
+  using PointVector  = std::vector<Point>;
+
+  using CurveGeometry = NURBSPatchGeometry<1, 2>;
+
+  // To keep track of which point belongs to which edge or node we will use a pointMap
+  struct IntersectionPoint {
+    Point point;
+    bool alreadyVisited = false;
+
+    explicit IntersectionPoint(const ClipperPoint& _clipperPoint) : point({_clipperPoint.x, _clipperPoint.y}){};
+  };
+
+  using IntersectionPointMap    = std::map<int, std::vector<IntersectionPoint>>;
+  using IntersectionPointMapPtr = std::shared_ptr<std::map<int, std::vector<IntersectionPoint>>>;
+
+  // Helper functions
+  double distance(const ClipperPoint p1, const ClipperPoint p2) {
+    return std::sqrt(std::pow(p1.x - p2.x, 2) + std::pow(p1.y - p2.y, 2));
+  }
+
+  bool approxSamePoint(const Point& a, const ClipperPoint& b, const double tol = tolerance) {
+    return Dune::FloatCmp::eq(a, Point{b.x, b.y}, tol);
+  }
+
+  bool approxSamePoint(Point a, Point b, const double tol = tolerance) { return Dune::FloatCmp::eq(a, b, tol); }
+
+  bool approxSamePoint(ClipperPoint a, ClipperPoint b, const double tol = tolerance) {
+    return Dune::FloatCmp::eq(distance(a, b), 0.0, tol);
+  }
+
+  int getEdgeOrientation(const int edge) { return (edge == 0 || edge == 2) ? 0 : 1; }
+
+  int iPlusX(int i, int x) { return (i + x) % 4; }
+
+  std::array<Point, 2> curveStartEndPoint(const CurveGeometry& curve) {
+    std::array<int, 2> indices{0, static_cast<int>(curve.patchData_->controlPoints.size()[0]) - 1};
+    return {curve.patchData_->controlPoints.get({indices[0]}).p, curve.patchData_->controlPoints.get({indices[1]}).p};
+  }
+
+  PointVector getElementCorners(const auto& element) {
+    auto geo = element.geometry();
+
+    std::vector<Point> corners(4);
+    corners[0] = geo.corner(0);
+    corners[1] = geo.corner(1);
+    corners[2] = geo.corner(3);  // see dune book page 127 Figure 5.12
+    corners[3] = geo.corner(2);
+
+    return corners;
+  }
+
+  Clipper2Lib::PathD getElementEdges(const auto& element) {
+    auto corners = getElementCorners(element);
+    Clipper2Lib::PathD edges;
```

Comment:
You should call the function below to not duplicate code.

```cpp
  Clipper2Lib::PathD getElementEdges(const PointVector& corners) {
    Clipper2Lib::PathD edges;
    for (int i = 0; i < 4; ++i)
      edges.emplace_back(corners[i][0], corners[i][1]);
    return edges;
  }

  Clipper2Lib::PathD getElementEdges(const auto& element) {
    auto corners = getElementCorners(element);
    return getElementEdges(corners);
  }
```

and maybe rename the cone with the corners to `getElementEdgesFromElementCorners`